contracts/MarginCalculator.sol

```diff
@@ -756,7 +916,87 @@ contract MarginCalculator is Ownable {
+     *
+     *
+     *                  starting price + (ending price - starting price) * auction elapsed time
+     * then price = --------------------------------------------------------------------------
+     *                                      auction time
+     *
+     *
+     * @param _vaultCollateral vault collateral amount
+     * @param _vaultDebt vault short amount
+     * @param _cv option cash value
+     * @param _spotPrice option underlying asset price (in USDC)
+     * @param _auctionStartingTime auction starting timestamp (_spotPrice timestamp from chainlink)
+     * @param _collateralDecimals collateral asset decimals
+     * @param _isPut otoken type, true for put, false for call option
+     * @return price of 1 debt otoken in collateral asset scaled by collateral decimals
+     */
+    function _price(
+        FPI.FixedPointInt memory _vaultCollateral,
+        FPI.FixedPointInt memory _vaultDebt,
+        FPI.FixedPointInt memory _cv,
+        FPI.FixedPointInt memory _spotPrice,
+        uint256 _auctionStartingTime,
+        uint256 _collateralDecimals,
+        bool _isPut
+    ) internal view returns (uint256) {
+        // price of 1 repaid otoken in collateral asset, scaled to collateral decimals and explicitly rounded down
+        FPI.FixedPointInt memory price;
+        // auction ending price
+        FPI.FixedPointInt memory endingPrice = _vaultCollateral.div(_vaultDebt);
+
+        // auction elapsed time
+        uint256 auctionElapsedTime = now.sub(_auctionStartingTime);
+
+        // if auction ended, return ending price
+        if (auctionElapsedTime >= AUCTION_TIME) {
+            price = endingPrice;
+        } else {
+            // starting price
+            FPI.FixedPointInt memory startingPrice;
+
+            {
+                // store oracle deviation in a FixedPointInt (already scaled by 1e27)
+                FPI.FixedPointInt memory fixedOracleDeviation = FPI.fromScaledUint(oracleDeviation, SCALING_FACTOR);
+
+                if (_isPut) {
+                    startingPrice = FPI.max(_cv.sub(fixedOracleDeviation.mul(_spotPrice)), ZERO);
+                } else {
+                    startingPrice = FPI.max(_cv.sub(fixedOracleDeviation.mul(_spotPrice)), ZERO).div(_spotPrice);
+                }
+            }
+
+            // store auctionElapsedTime in a FixedPointInt scaled by 1e27
+            FPI.FixedPointInt memory auctionElapsedTime = FPI.fromScaledUint(auctionElapsedTime, 18);
+            // store AUCTION_TIME in a FixedPointInt (already scaled by 1e27)
+            FPI.FixedPointInt memory auctionTime = FPI.fromScaledUint(AUCTION_TIME, 18);
+
+            // calculate price of 1 repaid otoken, scaled by the collateral decimals, expilictly rounded down
+            price = startingPrice.add((endingPrice.sub(startingPrice)).mul(auctionElapsedTime).div(auctionTime));
+        }
+
+        return price.toScaledUint(_collateralDecimals, true);
```

Comment:
I think we will need to scales this price variable by another 1e8 or 1e18, instead of just collateral decimal? Otherwise if the collateral asset has few decimals (for example of collateral is eth but have 0 decimals), 1 otoken will always worth 0. (price will always be 0).

Reply:
Can u explain more ? "(for example of collateral is eth but have 0 decimals)" ?